### DISCUSSION TITLE
Way to asynchronously run OS.execute()?

### DISCUSSION THREAD

**Discussion by ghost** [godotengine/godot-proposals#8871] [General Discussions]
I would like a way to read from standard output, and write to standard input, and do so with a child process I executed, all asynchronously. Can all of this be technically done already or should I create a pull request? I'm relatively new to Godot and don't know much about what it can or can't do without a bit of research, and it is simpler to just ask and get a simple "yes" or "no". I have written [my own C++ library](https://github.com/time-killer-games/xproc) which does all of these things, so I am more than happy to create a pull request at some point if I absolutely need to.

According to [this](https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-execute) the OS.execute() function blocks the main thread and according to [this](https://docs.godotengine.org/en/stable/tutorials/performance/using_multiple_threads.html) multithreading is possible in Godot, so on the surface, while putting one and one together, it would seem what I am asking is already possible in Godot. Is this true, and if so, does this apply to Godot 4 only or should it work in Godot 3 as well?

Also afaik there is a way to read standard output of a child process, but I saw nothing in the docs about writing to standard input of a child process, and in which case it seems that much will need a pull request from me one way or another if I am not sadly mistaken. Is this correct as well?

This is intended to make chess games with chess engine AI's like stockfish chess, for example, and among others, all of which are used as a IPC command line interface. (In my specific use case, this is why I need these features, but other use cases also exist obviously).

Thank you!

**Comment by AThousandShips** [comment]
Have you tried running it in a thread? See [here](https://docs.godotengine.org/en/stable/tutorials/performance/using_multiple_threads.html)

This is for proposing features, for support questions and help please first turn to the other [community channels](https://godotengine.org/community) 🙂

**Reply by ghost** [reply]
In a way this was half-proposing a feature, because if it was not possible already I was suggesting I added it in a pull request. There is also the part in this question where I was asking whether it is possible to read stdout and write stdin of the child process, which is a separate question from whether the process execution itself can be done in a thread. Since OS.execute is blocking, it would make sense even with multi-threading that you wouldn't be able to read stdout until after execution is over, which is something I would like to add support for to read stdout async, if not already possible. As well as write to the child process's stdin.

**Comment by bruvzg** [comment]
os.execute will block the current thread, so you can run it in a thread. It can read output, but it won't return until the process is finished (and will set output string only after finishing).

**Reply by Calinou** [reply]
https://github.com/godotengine/godot-proposals/issues/216 could also help address this with non-blocking `OS.execute()`.

In the meantime, you could have your process write to a file that is flushed every time a line is written, and read that file from Godot every so often. File-based IPC is hardly efficient but it may be good enough for some use cases.

**Reply by ghost** [reply]
Yeah, setting the output string only when execution was finished will be a problem for me. If I added a way to read from standard output and write to standard input of the child process all async, would that be acceptable for a pull request? I was thinking we could discuss how to go about the implementation in this ticket. For example, writing to a file might be very slow, as previously mentioned, and is not the best solution for every use case.

**Reply by ghost** [reply]
After seeing [#216 ](https://github.com/godotengine/godot-proposals/issues/216) that definitely looks a lot better that what we currently have, and I'm all for it. Though that ticket still doesn't address my thought on support for writing to Standard Input, how does everyone feel about adding an API for that as well?
